internals/secrethub/audit.go

```diff
@@ -150,24 +337,188 @@ func (cmd *AuditCommand) iterAndAuditTable() (secrethub.AuditEventIterator, audi
 	return nil, nil, ErrNoValidRepoOrSecretPath
 }
 
+// pager is an io.WriteCloser that returns errPagerClosed if it has been closed.
+type pager io.WriteCloser
+
+// newPaginatedWriter runs the terminal pager configured in the OS environment
+// and returns a writer that is piped to the standard input of the pager command.
+func newPaginatedWriter(outputWriter io.Writer) (pager, error) {
+	pager, err := pagerCommand()
+	if err != nil {
+		return nil, err
+	}
+
+	cmd := exec.Command(pager)
+
+	writer, err := cmd.StdinPipe()
+	if err != nil {
+		return nil, err
+	}
+
+	cmd.Stdout = outputWriter
+	cmd.Stderr = os.Stderr
+
+	err = cmd.Start()
+	if err != nil {
+		return nil, err
+	}
+	done := make(chan struct{}, 1)
+	go func() {
+		_ = cmd.Wait()
+		done <- struct{}{}
+	}()
+	return &paginatedWriter{writer: writer, cmd: cmd, done: done}, nil
+}
+
+// paginatedWriter is a writer that is piped to a terminal pager command.
+type paginatedWriter struct {
+	writer io.WriteCloser
+	cmd    *exec.Cmd
+	done   <-chan struct{}
+	closed bool
+}
+
+// Write pipes the data to the terminal pager.
+// It returns errPagerClosed if the terminal pager has been closed.
+func (p *paginatedWriter) Write(data []byte) (n int, err error) {
+	if p.isClosed() {
+		return 0, errPagerClosed
+	}
+	return p.writer.Write(data)
+}
+
+// Close closes the writer to the terminal pager and waits for the terminal pager to close.
+func (p *paginatedWriter) Close() error {
+	err := p.writer.Close()
+	if err != nil {
+		return err
+	}
+	if p.closed {
+		return nil
+	}
+	err = p.cmd.Process.Signal(syscall.SIGINT)
+	if err != nil {
+		err = p.cmd.Process.Kill()
+		if err != nil {
+			return err
+		}
+	}
+	<-p.done
+	return nil
+}
+
+// isClosed checks if the terminal pager process has been stopped.
+func (p *paginatedWriter) isClosed() bool {
+	if p.closed {
+		return true
+	}
+	select {
+	case <-p.done:
+		p.closed = true
+		return true
+	default:
+		return false
+	}
+}
+
+// pagerCommand returns the name of the terminal pager configured in the OS environment ($PAGER).
+// If no pager is configured it falls back to "less" than "more", returning an error if neither are available.
+func pagerCommand() (string, error) {
+	var pager string
+	var err error
+
+	pager, err = exec.LookPath(os.ExpandEnv(pagerEnvvar))
+	if err == nil {
+		return pager, nil
+	}
```

Comment:
Nice, thanks for the tip :+1:

Reply:
For this to work, you also have to remove the `var` statements at the beginning of the functions and use `:=` instead of `=`. Only then the variables are tightly scoped.

```golang
if pager, err := exec.LookPath(os.ExpandEnv(pagerEnvvar)); err == nil {
	return pager, nil
}
// This does not work because pager is out of scope here:
fmt.Println(pager)
```

Reply:
Missed that. Thanks for pointing it out!